mesonbuild/environment.py

```diff
@@ -1092,99 +1090,88 @@ def get_compiler_system_dirs(self):
         out = out.split('\n')[index].lstrip('libraries: =').split(':')
         return [os.path.normpath(p) for p in out]
 
+    def need_exe_wrapper(self, for_machine: MachineChoice = MachineChoice.HOST):
+        value = self.properties[for_machine].get('needs_exe_wrapper', None)
+        if value is not None:
+            return value
+        return not self.machines[for_machine].can_run()
+
     def get_exe_wrapper(self):
-        if not self.cross_info.need_exe_wrapper():
+        if not self.need_exe_wrapper():
             from .dependencies import EmptyExternalProgram
             return EmptyExternalProgram()
         return self.exe_wrapper
 
-
-class CrossBuildInfo:
-    def __init__(self, filename):
-        self.config = {'properties': {}}
-        self.parse_datafile(filename)
-        if 'host_machine' not in self.config and 'target_machine' not in self.config:
-            raise mesonlib.MesonException('Cross info file must have either host or a target machine.')
-        if 'host_machine' in self.config and 'binaries' not in self.config:
-            raise mesonlib.MesonException('Cross file with "host_machine" is missing "binaries".')
-
-    def ok_type(self, i):
-        return isinstance(i, (str, int, bool))
-
-    def parse_datafile(self, filename):
+class MesonConfigFile:
+    @classmethod
+    def parse_datafile(cls, filename):
         config = configparser.ConfigParser()
         try:
             with open(filename, 'r') as f:
                 config.read_file(f, filename)
         except FileNotFoundError:
             raise EnvironmentException('File not found: %s.' % filename)
+        return cls.from_config_parser(config)
+
+    @classmethod
+    def from_config_parser(cls, parser: configparser.ConfigParser):
+        out = {}
         # This is a bit hackish at the moment.
-        for s in config.sections():
-            self.config[s] = {}
-            for entry in config[s]:
-                value = config[s][entry]
+        for s in parser.sections():
+            section = {}
+            for entry in parser[s]:
+                value = parser[s][entry]
+                # Windows paths...
+                value = value.replace('\\', '\\\\')
```

Comment:
I copied that from your parser code, but perhaps we should move it so that it only applies to patches. We couldn't want to mess up escape sequences or whatever.

(But if so, let's do that in a separate PR. The tests are fine either way and this matches the current native file semantics.)

Reply:
We talked about this when I added that. the conclusion was, don't put escape sequences in your native file entires.